darknet_video.py

```diff
@@ -1,115 +1,103 @@
-                    pass
-        except Exception:
-            pass
-    #cap = cv2.VideoCapture(0)
-    cap = cv2.VideoCapture("test.mp4")
-    cap.set(3, 1280)
-    cap.set(4, 720)
-    out = cv2.VideoWriter(
-        "output.avi", cv2.VideoWriter_fourcc(*"MJPG"), 10.0,
-        (darknet.network_width(netMain), darknet.network_height(netMain)))
-    print("Starting the YOLO loop...")
+def main():
+    args = parser()
+    check_arguments_errors(args)
+    random.seed(3)  # deterministic bbox colors
 
-    # Create an image we reuse for each detect
-    darknet_image = darknet.make_image(darknet.network_width(netMain),
-                                    darknet.network_height(netMain),3)
-    while True:
-        prev_time = time.time()
-        ret, frame_read = cap.read()
-        frame_rgb = cv2.cvtColor(frame_read, cv2.COLOR_BGR2RGB)
-        frame_resized = cv2.resize(frame_rgb,
-                                   (darknet.network_width(netMain),
-                                    darknet.network_height(netMain)),
-                                   interpolation=cv2.INTER_LINEAR)
+    network, class_names, class_colors = darknet.load_network(
+        args.config_file,
+        args.data_file,
+        args.weights
+    )
+    # Darknet doesn't accept numpy images.
+    # Create one with image we reuse for each detect
+    width = darknet.network_width(network)
+    height = darknet.network_height(network)
+    darknet_image = darknet.make_image(width, height, 3)
 
-        darknet.copy_image_from_bytes(darknet_image,frame_resized.tobytes())
+    cap = cv2.VideoCapture(args.input)
+    video = set_video(args.input, args.out_filename, (width, height))
 
-        detections = darknet.detect_image(netMain, metaMain, darknet_image, thresh=0.25)
-        image = cvDrawBoxes(detections, frame_resized)
+    while cap.isOpened():
+        prev_time = time.time()
+        ret, frame = cap.read()
+        if not ret:
+            break
+        frame_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
+        frame_resized = cv2.resize(frame_rgb, (width, height),
+                                   interpolation=cv2.INTER_LINEAR)
+        darknet.copy_image_from_bytes(darknet_image, frame_resized.tobytes())
+        detections = darknet.detect_image(network, class_names, darknet_image, thresh=args.thresh)
+        image = darknet.draw_boxes(detections, frame_resized, class_colors)
         image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
-        print(1/(time.time()-prev_time))
-        cv2.imshow('Demo', image)
-        cv2.waitKey(3)
+        if args.out_filename is not None:
+            video.write(image)
+        fps = int(1/(time.time() - prev_time))
+        print("FPS: {}".format(fps))
+        darknet.print_detections(detections, args.ext_output)
+        if not args.dont_show:
+            cv2.imshow('Inference', image)
+            cv2.waitKey(fps)
     cap.release()
-    out.release()
+    video.release()
```

Comment:
Can we use 3 separate CPU-threads to process asychronously?
1. CPU-thread: VideoCapture + cvtColor + resize + copy_image_from_bytes
2. CPU-thread: darknet.detect_image
3. CPU-thread: draw_boxes + cvtColor + video.write + imshow + waitKey

Reply:
Good idea. I have to check that *point.2 CPU-thread: darknet.detect_image*, runs on GPU when possible, and python doesn't force CPU usage because bad multithread implementation.

IMHO, *point 1* and *point 3* can run always on CPU, because people rarely use opencv compiled with cuda. What your opinion?

Reply:
Yes, use CPU for points 1 and 2.

Since darknet.detect_image isn't async function, so creaste separate CPU-thread for it too.

So there should be 3 CPU-threads.

As in the C++ example - there are
*  1 default CPU-thread for Rendering(Show, Save to file, Send to web-browser)
and I create 2 additional CPU-threads: 
2. Inference (detection)
3. and VideoCapture+resizing